### PR TITLE
fix(gateway): return real usage for OpenAI-compatible chat completions

### LINKED ISSUE #49753: [Bug]: OpenAI-compatible custom providers lose streaming usage data, causing zero session tokens and unknown dashboard usage
### Bug type

Behavior bug (incorrect output/state without crash)

### Summary

OpenClaw drops token usage for OpenAI-compatible custom providers in the `openai-completions` streaming path, even when the upstream provider returns valid usage data. As a result:

  - session logs persist `usage.totalTokens = 0`
  - the dashboard shows `unknown/...`
  - `openclaw status --usage` shows no usable session/provider usage

  This does not appear to be a dashboard bug. The usage is already missing in the persisted session messages.

### Steps to reproduce

  1. Configure a custom provider using api: "openai-completions".
  2. Use that provider for an agent.
  3. Send a simple prompt, for example:

  openclaw agent --agent main --message "Reply with exactly OK."

  4. Check usage status:

  openclaw status --usage

  5. Inspect the latest session log for the assistant message.

### Expected behavior

If the upstream OpenAI-compatible provider returns valid usage fields, OpenClaw should persist them in the assistant message / agent metadata, and the dashboard / CLI should show actual token usage.

### Actual behavior

OpenClaw still persists zero usage for the assistant message, for example:

  "usage": {
    "input": 0,
    "output": 0,
    "cacheRead": 0,
    "cacheWrite": 0,
    "totalTokens": 0,
    "cost": {
      "input": 0,
      "output": 0,
      "cacheRead": 0,
      "cacheWrite": 0,
      "total": 0
    }
  }

  And the CLI shows:

  - session tokens as unknown/...
  - Usage: no provider usage available.

### OpenClaw version

2026.3.13

### Operating system

Ubuntu 24.4

### Install method

npm global

### Model

gpt-oss-120b, qwen3-235b-a22b-instruct-2507, qwen3-coder-30b-a3b-instruct, Qwen3-embedding-8b

### Provider / routing chain

custom-api-scaleway-ai

### Config file / key location

_No response_

### Additional provider/model setup details

_No response_

### Logs, screenshots, and evidence

```shell

```

### Impact and severity

_No response_

### Non-streaming request

  POST /chat/completions

  Response includes:

  "usage": {
    "prompt_tokens": 72,
    "completion_tokens": 8,
    "total_tokens": 80
  }

  ### Streaming request with usage enabled

  Request body includes:

  {
    "stream": true,
    "stream_options": {
      "include_usage": true
    }
  }

  {
    "choices": [],
    "usage": {
      "prompt_tokens": 72,
      "completion_tokens": 8,
      "total_tokens": 80
    }
  }

  So the provider is not the problem, and the dashboard is not the primary problem either. The usage is being lost inside OpenClaw before persistence/display.

  ## Additional Notes

  - Updating from 2026.3.7 to 2026.3.13 did not fix the issue.
  - A local workaround that injects stream_options.include_usage = true for the affected provider did not change the final persisted result.
  - This suggests the usage is likely being dropped deeper in the OpenAI-compatible streaming ingestion path, possibly around the final SSE chunk handling or propagation into lastAssistant / agentMeta.

  ## Suspected Area

  The issue appears to be in the openai-completions streaming path for OpenAI-compatible providers, especially handling of the final SSE chunk with:

  {
    "choices": [],
    "usage": { ... }
  }

  Potentially relevant area:

  - the imported streamSimple path from @mariozechner/pi-ai
  - usage propagation into final assistant message / agentMeta.lastCallUsage

  ## Impact

  This makes usage reporting effectively unusable for affected custom OpenAI-compatible providers:

  - dashboard usage is wrong / empty
  - CLI usage is wrong / empty
  - local session cost / usage aggregation is incomplete

  ## Suggested Fix Direction

  - Verify that the final chat.completion.chunk SSE event carrying usage is not discarded when choices is empty.
  - Ensure prompt_tokens, completion_tokens, and total_tokens are propagated into the final assistant message or agentMeta.lastCallUsage for openai-completions custom providers.

### PR DESCRIPTION
**Pull request by Lellansin**
## Summary

- Problem: `/v1/chat/completions` returned hardcoded zero usage in non-stream responses, and stream responses did not expose usage in an OpenAI-compatible way.
- Why it matters: OpenAI-compatible clients may depend on real usage values for telemetry, metering, assertions, or downstream compatibility behavior.
- What changed: the gateway now derives usage from `result.meta.agentMeta.usage`, returns it for non-stream responses, and emits a final usage chunk for stream responses when `stream_options.include_usage=true`.
- What changed (stream finalization semantics): streaming teardown now uses the same **finalize + usage gate** pattern as `openresponses-http.ts` (wait for the ingress command to settle when usage output is requested) instead of closing the SSE immediately from the prior `lifecycle` / `finally` paths. This is **adapter-level close ordering** so the optional usage chunk can precede `[DONE]`; it is **not** an HTTP-layer `setTimeout` removal relative to `main`.
- What changed (usage mapping): chat-completions usage **clamps non-negative components** before deriving `total_tokens`, and preserves upstream aggregate totals via **`max(componentTotal, aggregateTotal)`** when an aggregate total is present.
- What did NOT change (scope boundary): no auth/model-routing/request-validation/tool-behavior changes; OpenAI-compatible usage mapping keeps `prompt_tokens = input + cacheRead` (cache-write excluded intentionally).

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [x] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #38735
- Closes #49753
- Related PR
  - #46293
  - #38893
  - #24604
- [x] This PR fixes a bug or regression

## Root Cause (if applicable)

- Root cause: the OpenAI-compatible `/v1/chat/completions` adapter hardcoded zero usage for non-stream responses and did not implement a compatible usage-return path for streaming responses.
- Missing detection / guardrail: there was no targeted gateway coverage asserting that non-stream responses propagate real usage or that streaming usage finalization behaves correctly under lifecycle timing races.
- Contributing context (if known): prior attempts addressed parts of the behavior but did not land, so the compatibility gap remained in the adapter layer.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [ ] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `src/gateway/openai-http.test.ts` (endpoint contract) and `src/agents/usage.test.ts` (`toOpenAiChatCompletionsUsage` / normalization edge cases).
- Scenario the test should lock in: non-stream responses return real usage from `agentMeta.usage`; stream responses emit a final usage chunk only when `stream_options.include_usage=true`; normal streaming finalization is not blocked when usage output is not requested.
- Why this is the smallest reliable guardrail: the bug is in the gateway compatibility adapter and response-shaping/finalization logic, so targeted endpoint tests cover the exact contract with minimal surface area.
- Existing test that already covers this (if any): existing streaming/non-stream endpoint tests covered basic response flow, but not usage propagation or finalize race handling.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes

- Non-stream `/v1/chat/completions` responses now return real `usage` values instead of hardcoded zeros when upstream usage metadata exists.
- Stream `/v1/chat/completions` responses now emit a final usage chunk when `stream_options.include_usage=true`.
- Default stream behavior remains compatible for clients that do not request usage.

## Diagram (if applicable)

Before:
```mermaid
flowchart LR
    A["User request"] --> B["/v1/chat/completions adapter"]
    B --> C["Hardcoded usage = 0 / no streaming usage path"]
```

After:
```mermaid
flowchart LR
    A["User request"] --> B["/v1/chat/completions adapter"]
    B --> C["Derive usage from meta.agentMeta.usage<br/>(normalizeUsage → toOpenAiChatCompletionsUsage)"]
    C --> D["Non-stream: usage in JSON body"]
    C --> E["Stream: finalize gate + optional usage chunk<br/>when stream_options.include_usage = true"]
```

## Security Impact (required)

- New permissions/capabilities? (`No`)
- Secrets/tokens handling changed? (`No`)
- New/changed network calls? (`No`)
- Command/tool execution surface changed? (`No`)
- Data access scope changed? (`No`)
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: macOS
- Runtime/container: local Node.js / gateway build
- Model/provider: OpenClaw gateway OpenAI-compatible adapter, validated against agent runs that populate `result.meta.agentMeta.usage`
- Integration/channel (if any): Gateway HTTP `/v1/chat/completions`
- Relevant config (redacted): optional `OPENCLAW_DEBUG_OPENAI_USAGE=1`, gateway OpenAI-compatible endpoint enabled

### Steps

1. Start the gateway with the OpenAI-compatible endpoint enabled.
2. Send a non-stream POST to `/v1/chat/completions` and inspect the `usage` field.
3. Send a stream POST to `/v1/chat/completions` with `stream_options.include_usage=true` and inspect the final SSE usage chunk.

### Expected

- Non-stream responses return real usage values when available.
- Streaming responses emit usage only when explicitly requested.
- Normal streaming finalization still closes promptly when usage output is not requested.

### Actual

- Non-stream responses map real usage from `agentMeta.usage`.
- Streaming responses emit a final usage chunk when requested.
- Finalization race handling is covered so early lifecycle end does not hang usage-enabled streams.

## Evidence

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)

- Verified scenarios: local endpoint tests for non-stream usage mapping, stream usage chunk emission, lifecycle-end-before-usage race handling, and default stream finalization without usage gating.
- Edge cases checked: cache-read/cache-write usage mapping, explicit total override, stream finalize ordering, and default stream behavior when `include_usage` is absent.
- What you did **not** verify: full provider matrix across every upstream transport and external SDK client behavior outside the gateway test surface.

## Review Conversations

- [ ] I replied to or resolved every bot review conversation I addressed in this PR.
- [ ] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes`)
- Config/env changes? (`No`)
- Migration needed? (`No`)
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: upstream usage payloads may vary by provider shape or omit some fields.
  - Mitigation: usage is derived from normalized `agentMeta.usage` fields with existing fallback behavior.
- Risk: streaming finalization can regress if lifecycle timing changes.
  - Mitigation: added targeted tests for lifecycle-end-before-usage and for non-usage stream finalization behavior.

## Trade-offs / Deferred follow-up

- Current decision in this PR: keep `/v1/chat/completions` stream finalization semantics aligned with `openresponses-http.ts` (finalize request + final usage gate), instead of introducing a chat-completions-only fallback.
- Known trade-off: when `stream_options.include_usage=true`, completion latency can still depend on when the ingress command settles on very slow tail/cleanup paths.
- Why this scope boundary: this PR is focused on usage correctness and OpenAI-compatible usage payload behavior, and we want to avoid cross-endpoint semantic divergence in finalize behavior.
- Follow-up direction (shared adapter decision): evaluate a bounded-finalize strategy for both `openai-http` and `openresponses-http` together.

### Evidence links

- `src/gateway/openai-http.ts` — OpenAI-compatible `/v1/chat/completions` adapter: usage mapping, streaming finalize gate, optional usage chunk when `stream_options.include_usage=true`.
- `src/gateway/openai-http.test.ts` — Contract tests for non-stream usage, streaming usage chunk, finalize/lifecycle behavior, and default stream path without usage.
- `src/agents/usage.ts` — `normalizeUsage` + `toOpenAiChatCompletionsUsage` (OpenAI-style `prompt_tokens` / `completion_tokens` / `total_tokens`, including aggregate vs component totals).
- `src/agents/usage.test.ts` — Unit tests for chat-completions usage mapping edge cases.

**Reference (not modified in this PR):** `src/gateway/openresponses-http.ts` — finalize / usage-gating pattern this endpoint mirrors in comments and behavior.

### Related unresolved threads

- https://github.com/openclaw/openclaw/pull/62986#discussion_r3062088677
- https://github.com/openclaw/openclaw/pull/62986#discussion_r3062229527

### CODE REVIEW

File: src/gateway/openai-http.ts [hunk 1]
```diff
@@ -558,9 +753,68 @@ export async function handleOpenAiHttpRequest(
...
+  const requestFinalize = () => {
+    if (!finalizeRequested) {
+      finalizeRequested = true;
+    }
+    if (streamIncludeUsage && !finalUsage && !finalizeWithoutUsage && !usageFinalizeTimer) {
+      usageFinalizeTimer = setTimeout(() => {
+        usageFinalizeTimer = undefined;
+        finalizeWithoutUsage = true;
+        abortController.abort();
+        maybeFinalize();
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts]: Thanks, addressed in the latest update.

I removed the forced 1s close path (`finalizeWithoutUsage` / timer abort) from `src/gateway/openai-http.ts`, and aligned the stream finalize gate with `openresponses-http.ts`: for `stream_options.include_usage=true`, the stream now waits for both finalize request and final usage before emitting `[DONE]`.

I also updated the related test coverage in `src/gateway/openai-http.test.ts` to validate disconnect cleanup instead of timeout-based completion.

File: src/gateway/openai-http.ts [hunk 2]
```diff
@@ -421,6 +459,148 @@ function resolveAgentResponseText(result: unknown): string {
...
+  const usage = normalizeUsage(resolveRawAgentUsage(result));
+
+  // Keep the wire-format aligned with OpenAI chat completions:
+  // prompt_tokens includes prompt input plus cache-read reuse, but not cache-write overhead.
+  const input = usage?.input ?? 0;
+  const output = usage?.output ?? 0;
+  const cacheRead = usage?.cacheRead ?? 0;
+  const promptTokens = input + cacheRead;
+  const completionTokens = output;
+  const totalTokens = promptTokens + completionTokens;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts]: Thanks for calling this out.

This behavior is intentional for OpenAI-compat mapping in `/v1/chat/completions`: `prompt_tokens` is mapped as `input + cacheRead`, and `cacheWrite` is intentionally excluded in `src/gateway/openai-http.ts` to match expected OpenAI-style semantics on this wire surface.

`total_tokens` is derived as `prompt_tokens + completion_tokens` for consistency on the compatibility response. We still keep full internal usage metadata (including cache-write) in internal/debug paths; we’re only keeping the public compatibility fields aligned.

**Comment by Lellansin** [comment]
@codex review

File: src/gateway/openai-http.ts [hunk 3]
```diff
@@ -421,6 +458,148 @@ function resolveAgentResponseText(result: unknown): string {
...
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts]: Addressed in the latest branch update.

I updated `resolveChatCompletionUsage` to preserve aggregate `usage.total` only when no normalized component-level breakdown is available. This keeps OpenAI-compat semantics for component-based mapping while fixing under-reporting for aggregate-only metadata.

I also added focused coverage in `src/gateway/openai-http.test.ts` for:
- aggregate-only usage (`{ total: 123 }`) in non-stream and stream (`include_usage=true`),
- non-finite component fields with aggregate fallback,
- cacheWrite-only payloads (no aggregate fallback re-introduction).

File: src/gateway/openai-http.ts [hunk 4]
```diff
@@ -558,10 +752,49 @@ export async function handleOpenAiHttpRequest(
...
+  // Streaming finalize gate intentionally mirrors OpenResponses:
+  // lifecycle:end requests finalize, command result provides finalUsage.
+  // When include_usage=true, both signals must arrive before closing the stream.
+  // Flow reference: src/gateway/openresponses-http.ts:817-827,876-881
+  const maybeFinalize = () => {
+    if (closed || !finalizeRequested) {
+      return;
+    }
+    if (streamIncludeUsage && !finalUsage) {
+      return;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts:656]: Addressed in the latest branch update.

I added an explicit code comment in `src/gateway/openai-http.ts` next to the finalize gate to document why this path intentionally stays aligned with `openresponses-http.ts` and why it does not hang indefinitely in the current runner contract.

Key point: `agentCommandFromIngress` is expected to settle via normal completion, timeout, or abort. On success, the command path sets `finalUsage` and requests finalize; its `finally` path also emits `lifecycle:end`, so both signals reach `maybeFinalize()`. On error/timeout, the catch path sets zeroed fallback usage, emits `lifecycle:error`, and requests finalize, so the stream still closes with a final usage state. Client disconnect is the separate early-exit case: disconnect aborts the runner and the rejected promise exits early because the stream is already closed/aborted.

So the current behavior here is a parity decision with `openresponses-http.ts`, and the runner-level `lifecycle:error` path already handles the final-usage failure case before this HTTP adapter finalizes the stream.

File: src/gateway/openai-http.ts [hunk 5]
```diff
@@ -558,10 +759,53 @@ export async function handleOpenAiHttpRequest(
...
+  // so the gate unblocks once agentCommandFromIngress settles.
+  // Note: the agent timeout (default 48h) is not an HTTP-meaningful bound;
+  // real client-side protection comes from LB/TCP keepalive timeouts.
+  // Whether a shorter bounded fallback is needed is a shared adapter decision.
+  const maybeFinalize = () => {
+    if (closed || !finalizeRequested) {
+      return;
+    }
+    if (streamIncludeUsage && !finalUsage) {
+      return;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts:656]: Addressed in the latest branch update.

I added an explicit code comment in `src/gateway/openai-http.ts` next to the finalize gate to document why this path intentionally stays aligned with `openresponses-http.ts` and why it does not hang indefinitely in the current runner contract.

Key point: `agentCommandFromIngress` is expected to settle via normal completion, timeout, or abort. On success, the command path sets `finalUsage` and requests finalize; its `finally` path also emits `lifecycle:end`, so both signals reach `maybeFinalize()`. On error/timeout, the catch path sets zeroed fallback usage, emits `lifecycle:error`, and requests finalize, so the stream still closes with a final usage state. Client disconnect is the separate early-exit case: disconnect aborts the runner and the rejected promise exits early because the stream is already closed/aborted.

So the current behavior here is a parity decision with `openresponses-http.ts`, and the runner-level `lifecycle:error` path already handles the final-usage failure case before this HTTP adapter finalizes the stream.

File: src/gateway/openai-http.ts [hunk 6]
```diff
@@ -421,6 +458,66 @@ function resolveAgentResponseText(result: unknown): string {
...
+  const promptTokens = input + cacheRead;
+  const completionTokens = output;
+  const hasComponentBreakdown =
+    usage?.input !== undefined ||
+    usage?.output !== undefined ||
+    usage?.cacheRead !== undefined ||
+    usage?.cacheWrite !== undefined;
+  const totalTokens = hasComponentBreakdown
+    ? promptTokens + completionTokens
+    : Math.max(0, usage?.total ?? 0);
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts]: Addressed in commit 095c072ddc.

I switched `/v1/chat/completions` usage mapping to `toOpenAiChatCompletionsUsage` (`src/agents/usage.ts`), which preserves aggregate totals for partial breakdowns via `total_tokens = max(prompt+completion, aggregate total)` while keeping OpenAI-style `prompt_tokens = input + cacheRead`.

I also aligned endpoint tests in `src/gateway/openai-http.test.ts` for the `total=100` and `cacheWrite-only total=10` scenarios, and re-ran targeted tests:
`pnpm test src/agents/usage.test.ts src/gateway/openai-http.test.ts`.

File: src/gateway/openai-http.ts [hunk 7]
```diff
@@ -558,10 +759,64 @@ export async function handleOpenAiHttpRequest(
...
+  // marks the stream closed and aborts the runner, then the rejected promise
+  // returns early from catch because closed/aborted is already true.
+  // Whether adapters should also add a shorter HTTP-specific bounded fallback
+  // remains a shared adapter decision rather than a chat-completions-only change.
+  const maybeFinalize = () => {
+    if (closed || !finalizeRequested) {
+      return;
+    }
+    if (streamIncludeUsage && !finalUsage) {
+      return;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts:656]: Thanks for the follow-up — this is a fair concern.

I documented this trade-off directly in the PR body under **Trade-offs / Deferred follow-up** so it is not only captured in review comments.

For this PR, I kept behavior aligned with `openresponses-http.ts` on purpose (finalize request + final usage gate) and kept scope focused on usage correctness/OpenAI-compatible payload mapping.

Evidence in code/tests:
- `src/gateway/openai-http.ts` (include_usage finalize gate + command-settlement path)
- `src/gateway/openresponses-http.ts` (same finalize pattern)
- `src/agents/pi-embedded-subscribe.handlers.lifecycle.ts` (lifecycle `end` emitted before async flush completes)
- `src/agents/timeout.ts` (ingress default timeout behavior)
- `src/gateway/openai-http.test.ts` (lifecycle-before-usage, include_usage behavior, disconnect cleanup coverage)

I agree a shorter HTTP-bounded finalize path can be a valid improvement, but I propose to evaluate it as a shared adapter follow-up for both endpoints rather than a chat-completions-only divergence inside this PR.

File: src/gateway/openai-http.ts [hunk 8]
```diff
@@ -558,10 +657,77 @@ export async function handleOpenAiHttpRequest(
...
+    // - success: assigns real finalUsage, then requests finalize
+    // - error/timeout: assigns zeroed finalUsage via the lifecycle:error path,
+    //   then requests finalize
+    // - client disconnect: closes early via abort/watchClientDisconnect
+    //
+    // So this does not create a chat-completions-only infinite wait; a shorter
+    // HTTP-specific bounded fallback would be a shared adapter decision for both
+    // this endpoint and openresponses-http.ts.
+    if (streamIncludeUsage && !finalUsage) {
+      return;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/gateway/openai-http.ts:656]: Thanks for the follow-up — this is a fair concern.

I documented this trade-off directly in the PR body under **Trade-offs / Deferred follow-up** so it is not only captured in review comments.

For this PR, I kept behavior aligned with `openresponses-http.ts` on purpose (finalize request + final usage gate) and kept scope focused on usage correctness/OpenAI-compatible payload mapping.

Evidence in code/tests:
- `src/gateway/openai-http.ts` (include_usage finalize gate + command-settlement path)
- `src/gateway/openresponses-http.ts` (same finalize pattern)
- `src/agents/pi-embedded-subscribe.handlers.lifecycle.ts` (lifecycle `end` emitted before async flush completes)
- `src/agents/timeout.ts` (ingress default timeout behavior)
- `src/gateway/openai-http.test.ts` (lifecycle-before-usage, include_usage behavior, disconnect cleanup coverage)

I agree a shorter HTTP-bounded finalize path can be a valid improvement, but I propose to evaluate it as a shared adapter follow-up for both endpoints rather than a chat-completions-only divergence inside this PR.

File: src/agents/usage.ts [hunk 9]
```diff
@@ -143,6 +143,41 @@ export function normalizeUsage(raw?: UsageLike | null): NormalizedUsage | undefi
...
+  total_tokens: number;
+} {
+  const input = usage?.input ?? 0;
+  const output = usage?.output ?? 0;
+  const cacheRead = usage?.cacheRead ?? 0;
+  const promptTokens = input + cacheRead;
+  const completionTokens = output;
+  const componentTotal = promptTokens + completionTokens;
+  const aggregateRaw = usage?.total;
+  const aggregateTotal =
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/agents/usage.ts]: Addressed. I updated toOpenAiChatCompletionsUsage to clamp prompt/completion components before deriving componentTotal and total_tokens, so total_tokens can no longer fall below prompt_tokens + completion_tokens when malformed finite usage appears. I also added two regression tests in src/agents/usage.test.ts for {input:3, output:-5} => total_tokens:3 and {input:3, output:-5, total:7} => total_tokens:7 to preserve aggregate-total behavior. Verified with: pnpm test src/agents/usage.test.ts (27 passed). Base commit: 095c072ddc.

File: src/agents/usage.ts [hunk 10]
```diff
@@ -143,6 +143,41 @@ export function normalizeUsage(raw?: UsageLike | null): NormalizedUsage | undefi
...
+  prompt_tokens: number;
+  completion_tokens: number;
+  total_tokens: number;
+} {
+  const input = usage?.input ?? 0;
+  const output = usage?.output ?? 0;
+  const cacheRead = usage?.cacheRead ?? 0;
+  const promptTokens = Math.max(0, input + cacheRead);
+  const completionTokens = Math.max(0, output);
+  const componentTotal = promptTokens + completionTokens;
```
[the remark this answers is not in the document]
Comment (Lellansin) [reply src/agents/usage.ts:165]: Thanks for the flag — the underlying issue (finite but negative `cacheRead` / `cacheWrite` skewing derived totals) is real.

I’m **not** going to special-case this only on `/v1/chat/completions` right now. Today the same raw-ish usage shape is consumed elsewhere **without** clamping cache fields to `>= 0` first:

- **OpenResponses** (`openresponses-http.ts` → `toUsage`): `cacheRead` / `cacheWrite` participate in the default `total = input + output + cacheRead + cacheWrite` path without a prior non-negative clamp; only the emitted `input_tokens` / `output_tokens` / `total_tokens` get `Math.max(0, …)` on those specific fields.
- **OpenAI WebSocket** (`openai-ws-message-conversion.ts` → `buildAssistantMessageFromResponse`): usage comes from `normalizeUsage(response.usage)` and is passed through `buildUsageWithNoCost` with **no** extra cache clamp; the fallback total also sums `input + output + cacheRead + cacheWrite` as-is.

So fixing this **only** inside the chat-completions mapper would make this endpoint behave differently from those paths for the same malformed provider metadata. If we harden this, it should be **shared** (e.g. clamp non-negative `cacheRead`/`cacheWrite` in `normalizeUsage`, or a single normalization helper used by OpenResponses + WS + compat), not a chat-completions-only special case.

**Review by obviyus — APPROVED** [review]
Verified the OpenAI-compatible chat completions usage fix path on the latest head. This keeps real usage in non-stream responses, emits the final usage chunk when include_usage is requested, and moves the terminal lifecycle signal behind the final flush so the compat stream no longer needs the truncating HTTP fallback.

During landing I rebased onto latest main, kept the changelog entry in the active Unreleased fixes block, and replaced the 2s HTTP abort patch with the root-cause lifecycle fix in pi-embedded subscribe. That addressed the current-head P1 about truncating slow-tail output after lifecycle end.

Local gate: oxfmt passed on the touched files. I also ran the targeted lifecycle/openai-http test lane, but the Vitest wrapper in this worktree did not return a clean exit, so I am not claiming a fully green local test run.

**Comment by obviyus** [comment]
Landed.

Maintainer commits during landing:
- [40138ef572de6cd702c059c19973964cc3df3c21](https://github.com/openclaw/openclaw/commit/40138ef572de6cd702c059c19973964cc3df3c21) fix(agents): emit lifecycle terminal events after flush
- [591a39a68bb2dcb28f265abffbed938aed2d8344](https://github.com/openclaw/openclaw/commit/591a39a68bb2dcb28f265abffbed938aed2d8344) fix: add OpenAI compat usage changelog (#62986) (thanks @Lellansin)
- [e964469d973d9669f06dfb1ef75732bc715fb4a5](https://github.com/openclaw/openclaw/commit/e964469d973d9669f06dfb1ef75732bc715fb4a5) fix(gateway): bound usage stream finalization

Merged as [2ccd1839f212d84dfb78f78e3699c384a12ee091](https://github.com/openclaw/openclaw/commit/2ccd1839f212d84dfb78f78e3699c384a12ee091).

Thanks @Lellansin.
